iceberg/src/main/scala/com/nvidia/spark/rapids/iceberg/GpuIcebergPartitioner.scala

```diff
@@ -0,0 +1,206 @@
+import org.apache.iceberg.spark.{GpuTypeToSparkType, SparkStructLike}
+import org.apache.iceberg.spark.functions.GpuBucketExpression
+import org.apache.iceberg.types.Types
+
+import org.apache.spark.sql.catalyst.expressions.GenericRowWithSchema
+import org.apache.spark.sql.catalyst.expressions.NamedExpression.newExprId
+import org.apache.spark.sql.types.{DataType, StructType}
+import org.apache.spark.sql.vectorized.{ColumnarBatch, ColumnVector}
+
+/**
+ * A GPU based Iceberg partitioner that partitions the input columnar batch into multiple
+ * columnar batches based on the given partition spec.
+ *
+ * @param spec the iceberg partition spec
+ * @param dataSparkType the spark schema of the input data
+ */
+class GpuIcebergPartitioner(val spec: PartitionSpec,
+  val dataSparkType: StructType) {
+  require(spec.isPartitioned, "Should not create a partitioner for unpartitioned table")
+  private val inputSchema: Schema = spec.schema()
+  private val sparkType: Array[DataType] = dataSparkType.fields.map(_.dataType)
+  private val partitionSparkType: StructType = GpuTypeToSparkType.toSparkType(spec.partitionType())
+
+  private val partitionExprs: Seq[GpuExpression] = spec.fields().asScala.map(getPartitionExpr).toSeq
+
+  private val keyColNum: Int = spec.fields().size()
+  private val keyColIndices: Array[Int] = (0 until keyColNum).toArray
+  private val keySortOrders: Array[OrderByArg] = (0 until keyColNum)
+    .map(OrderByArg.asc(_, true))
+    .toArray
+
+  /**
+   * Partition the `input` columnar batch using iceberg's partition spec.
+   * <br/>
+   * This method takes the ownership of the input columnar batch, and it should not be used after
+   * this call.
+   */
+  def partition(input: ColumnarBatch): Seq[ColumnarBatchWithPartition] = {
+    if (input.numRows() == 0) {
+      return Seq.empty
+    }
+
+    val numRows = input.numRows()
+
+    val spillableInput = SpillableColumnarBatch(input, ACTIVE_ON_DECK_PRIORITY)
+
+    val (partitionKeys, partitions) = withRetryNoSplit(spillableInput) { scb =>
+      val parts = withResource(scb.getColumnarBatch()) { inputBatch =>
+        partitionExprs.safeMap(_.columnarEval(inputBatch))
+      }
+      val keysTable = withResource(parts) { _ =>
+        val arr = new Array[CudfColumnVector](partitionExprs.size)
+        for (i <- partitionExprs.indices) {
+          arr(i) = parts(i).getBase
+        }
+        new Table(arr:_*)
+      }
+
+      val sortedKeyTableWithRowIdx = withResource(keysTable) { _ =>
+        withResource(Scalar.fromInt(0)) { zero =>
+          withResource(CudfColumnVector.sequence(zero, numRows)) { rowIdxCol =>
+            val totalColCount = keysTable.getNumberOfColumns + 1
+            val allCols = new Array[CudfColumnVector](totalColCount)
+
+            for (i <- 0 until keysTable.getNumberOfColumns) {
+              allCols(i) = keysTable.getColumn(i)
+            }
+            allCols(keysTable.getNumberOfColumns) = rowIdxCol
```

Comment:
If we generate a row index column, then it's no need to sort the keys table and data table.
Table.`groupBy` with the row index column and then `contiguousSplitGroupsAndGenUniqKeys` on keys columns, we can get the gather map for data table. Is it required to return sorted groups?

Reply:
Please ignore, we need the sort for the following `upper_bound`